renaissance-harness/src/main/java/org/renaissance/harness/ExecutionDriver.java

```diff
@@ -0,0 +1,141 @@
+import org.renaissance.ExecutionPolicy;
+import org.renaissance.core.BenchmarkInfo;
+
+/**
+ * Benchmark execution driver. Captures the sequence of actions performed
+ * during benchmark execution (i.e., calling the sequencing methods on a
+ * benchmark and issuing notifications to event listeners) and maintains
+ * execution context.
+ */
+final class ExecutionDriver implements BenchmarkContext {
+
+  private final BenchmarkInfo benchInfo;
+
+  private final Config config;
+
+
+  public ExecutionDriver(
+    final BenchmarkInfo benchInfo, final Config config
+  ) {
+    this.benchInfo = benchInfo;
+    this.config = config;
+  }
+
+
+  public final void executeBenchmark(
+    Benchmark benchmark, EventDispatcher dispatcher
+  ) throws ValidationException {
+    final ExecutionPolicy policy = config.policyFactory.create(config, benchInfo);
+    final String configuration = config.configuration;
+
+    benchmark.setUpBeforeAll(this);
+
+    try {
+      dispatcher.notifyAfterBenchmarkSetUp(benchInfo.name());
+
+      try {
+        int operationIndex = 0;
+
+        do {
+          printStartInfo(operationIndex, benchInfo, configuration);
+
+          final long durationNanos = executeOperation(operationIndex, benchInfo.name(), benchmark,
+            dispatcher, policy.isLastOperation()
+          );
+
+          printEndInfo(operationIndex, benchInfo, configuration, durationNanos);
+          policy.registerOperation(operationIndex, durationNanos);
+
+          operationIndex++;
+        } while (policy.keepExecuting());
+
+      } finally {
+        dispatcher.notifyBeforeBenchmarkTearDown(benchInfo.name());
+      }
+
+    } finally {
+      benchmark.tearDownAfterAll(this);
+    }
+  }
+
+
+  private long executeOperation(
+    final int opIndex, final String benchName, final Benchmark bench,
+    final EventDispatcher dispatcher, final boolean isLast
+  ) throws ValidationException {
+    //
+    // Call the benchmark and notify listeners before the measured operation.
+    // Nothing should go between the measured operation call and the timing
+    // calls. If everything goes well, we notify the listeners and call the
+    // benchmark after the measured operation and try to validate the result.
+    // If the result is valid, we notify listeners about basic result metrics
+    // and return the duration of the measured operation, otherwise an
+    // exception is thrown and the method terminates prematurely.
+    //
+    final long unixTsBefore = System.currentTimeMillis();
+    bench.beforeIteration(this);
+    dispatcher.notifyAfterOperationSetUp(benchName, opIndex, isLast);
+
+    final long startNanos = System.nanoTime();
+    final BenchmarkResult result = bench.runIteration(this);
+    final long durationNanos = System.nanoTime() - startNanos;
+
+    dispatcher.notifyBeforeOperationTearDown(benchName, opIndex, durationNanos);
+    bench.afterIteration(this);
+    final long unixTsAfter = System.currentTimeMillis();
```

Comment:
One more thing came to my mind after playing with the example plugins. Imagine user registers two plugins wrapping `runIteration`, one for hardware counters (inner) and one for `CompilationMXBean` (outer as it is more coarse-grained). Then if the hardware-counter plugin stores the data in a similar way as the current `ResultWriter`, the MX bean results would be tainted due to extra work done by the first plugin.

What I mean is that we miss callback that would allow postponing data processing in the plugins after we are outside the *possibly still measuring something* zone.

Not that we can anticipate all the possible interferences between all possible plugins but this seems to be an obvious spot (I remember the issues when adding hardware counters to DaCapo).

Reply:
Would an event after operation tear down help with this? Instead of `OperationTearDownListener`, we could have `BeforeOperationTearDownListener` and `AfterOperationTearDownListener`, with the latter providing a method with a callback parameter, which could be (optionally) used to publish measurement results.

This way a plugin could do quick things in the `beforeOperationTearDown` method (no parameter to provide results) and do other activities in the `afterOperationTearDown` method, optionally publishing measurement results.

A bit unrelated, but now that I think about it: it seems that `BenchmarkResultListener` is not a good name, because BenchmarkResult is the result returned by Benchmark), so maybe it should be renamed to MeasurementResultConsumer. In light of this, instead of `AfterOperationTearDownListener`, we could have a `MeasurementResultSupplier`, but the naming can be solved later.

So would this help? Of course, plugins could always screw things up in arbitrary fashion, but we assume responsible plugin writers (us :-)

Reply:
That sounds good and it also solves the problem how to push custom measurements from plugins to CSV/JSON.

Reply:
I have added initial support for publishing custom results. With the following code added to ExampleOne, it even produces additional results in the JSON output:
```java
@Override
public void onMeasurementResultsRequested(String benchmark, Plugin.MeasurementResultDispatcher dispatcher) {
    System.err.println(PREFIX + "measurements requested");
    dispatcher.publishMeasurementResult("dummy-one", 42);
}
```

Reply:
Now that I think about it, maybe we could use the existing MeasurementResultListener interface as the dispatcher parameter. The reason I defined a new interfaces was that I did not need the benchmark name parameter, but maybe that's not such a big deal and I would consider having one less interface a plus. What do you think?

Reply:
The current solution is perhaps a bit cleaner but reducing amount of things to learn/remember is always better :-).

Reply:
My sentiment exactly. Nobody says that a listener cannot be a composite which just forwards the notification to its parts, which is exactly what happens here. I think one less interface is worth it, so I've made the change :-)

BTW, plugins can now take command line parameters, if they implement a `(String... args)` constructor. If they don't implement it, the default is used. So the following command  is now possible:
```
java -jar target/renaissance-gpl-0.10.0.jar \
  --json output.json \
  --plugin 'plugins/one.jar!plugin.ExampleOne' \
  --plugin 'plugins/two.jar!plugin.ExampleTwo' \
  --with-arg first-argument \
  --with-arg second-argument \
  dummy-empty -r 3
```

I know that we can pass arguments through environment variables or system properties, but this removes the need to come up with property names :-)